lib/OpenQA/WebAPI/Controller/API/V1/Job.pm

```diff
@@ -388,28 +388,46 @@ sub update_status {
         my $err = 'Got status update for non-existing job: ' . $job_id;
         OpenQA::Utils::log_info($err);
         $self->render(json => {error => $err}, status => 400);
-        return;
+        return undef;
     }
 
-    if (!exists $status->{worker_id}) {
-        my $err = 'Got status update for job ' . $job_id . ' but does not contain a worker id!';
+    my $worker_id = $status->{worker_id};
+    if (!defined $worker_id) {
+        my $err = "Got status update for job $job_id but does not contain a worker id!";
         OpenQA::Utils::log_info($err);
         $self->render(json => {error => $err}, status => 400);
-        return;
+        return undef;
     }
 
+    # find worker
+    # - use either worker which current job is the one this status update is about
+    # - or use the assigned worker if the job is still running and that worker is not
+    #   known to run another job
     my $worker = $job->worker;
-    if (!$worker || $worker->id != $status->{worker_id}) {
-        my $worker_id  = $worker ? $worker->id : 'no updates anymore';
-        my $job_status = $job->status_info;
+    my $use_assigned_worker;
+    if (!$worker && !defined $job->t_finished) {
+        if (my $assigned_worker = $job->assigned_worker) {
+            $worker              = $assigned_worker;
+            $use_assigned_worker = 1;
+        }
+        else {
+            log_error("Wtf, not even a worker assigned");
```

Comment:
How about

```perl
my $job_status = $job->status_info;
log_error("Job is $job_status with no worker assigned");
```

Reply:
Oh, this leftover shouldn't have made it into the commit :-)

Reply:
This condition actually shows an API miss-usage. So I made it a `log_info` and return a decent error message to the user.